[@mantine/nprogress] Init

jerebtw opened the pull request:
This is a new version of my [old PR](https://github.com/mantinedev/mantine/pull/769) (Complete rewrite)

- [X] Better prop names
- [X] Demos
- [X] Docs

New PR of [old PR](https://github.com/mantinedev/mantine/pull/1218) but based on the master branch

jerebtw commented:
@rtivital can you please review this PR (I'm not really good at writing documents)

jerebtw commented:
Currently there is still one problem im not sure how to solve and that is when you want to restart the nprogress directly after it faded out because then it will not be visible for x ms (I will look in to that problem later this day)

rtivital commented on src/mantine-demos/src/Demos.story.tsx:
```diff
@@ -100,6 +100,7 @@ attachDemos(storiesOf('Styles', module), demos.CreateStylesDemos);
 attachDemos(storiesOf('Theme', module), demos.ThemeDemos);
 attachDemos(storiesOf('Spotlight', module), demos.SpotlightDemos);
 attachDemos(storiesOf('Form', module), demos.FormDemos);
+attachDemos(storiesOf('NProgress', module), demos.NProgressDemos);
```
This file will be removed or already removed, demos should be added other way to avoid performance issues. See Tabs demos for reference.

jerebtw replied:
Is that change also on the master branch? Because this is based on the master branch

rtivital replied:
No, this PR is for v5 branch, and it must use it as a base.

jerebtw replied:
Okay then I will rebase it later

rtivital commented on src/mantine-nprogress/src/NProgress.tsx:
```diff
@@ -0,0 +1,126 @@
+  unmountTime?: number;
+
+  /** Unmount transition duration in ms after progress is 100% */
+  unmountDuration?: number;
```
Rename to `exitTransitionDuration`

rtivital commented on src/mantine-nprogress/src/NProgress.tsx:
```diff
@@ -0,0 +1,126 @@
+  barTransitionDuration?: number;
+
+  /** The time when the component should be unmounted after progress is 100% */
+  unmountTime?: number;
```
Rename to `exitTimeout`

rtivital commented on src/mantine-nprogress/src/NProgress.tsx:
```diff
@@ -0,0 +1,126 @@
+    <OptionalPortal withinPortal={withinPortal} zIndex={zIndex}>
+      <Transition
+        mounted={mounted}
+        duration={1}
```
Why duration is 1?

jerebtw replied:
Thx forgot to change that

rtivital commented on src/mantine-nprogress/src/NProgress.tsx:
```diff
@@ -0,0 +1,126 @@
+  onFinish?: () => void;
+
+  /** Interval time in ms */
+  intervalTime?: number;
```
Almost all props that involve timeouts are named poorly – it is not clear what they do, please review them and add better names and descriptions.

jerebtw commented:
I rebased the PR and updated the prop names

> Currently there is still one problem im not sure how to solve and that is when you want to restart the nprogress directly after it faded out because then it will not be visible for x ms (I will look in to that problem later this day)

This is now also fixed

@rtivital Can you check the PR

rtivital commented on src/mantine-nprogress/src/NProgress.tsx:
```diff
@@ -0,0 +1,122 @@
+  const add = (value: number) => setProgress((c) => Math.min(c + value, 100));
+  const decrease = (value: number) => setProgress((c) => Math.max(c - value, 0));
+
+  const interval = useInterval(() => add(stepSize), stepIntervalTime);
```
I've researched https://github.com/rstacruz/nprogress/blob/e1a8b7fb6e059085df5f83c45d3c2308a147ca18/nprogress.js#L158-L177

Logic with for the interval should be different – instead of incrementing by a specific step, it uses different step values based on the current value. I think it will work much better that current implementation.

rtivital commented:
Thanks, great job!
